toolchain/check/context.cpp

```diff
@@ -216,6 +216,14 @@ auto Context::DiagnoseDuplicateName(SemIRLoc dup_def, SemIRLoc prev_def)
       .Emit();
 }
 
+auto Context::DiagnosePoisonedName(SemIRLoc dup_def) -> void {
+  CARBON_DIAGNOSTIC(
+      NameDeclPoisoned, Error,
+      "cannot declare this name in this scope since is was already used "
+      "without qualification in the context of this scope");
+  emitter_->Build(dup_def, NameDeclPoisoned).Emit();
```

Comment:
Thanks Jon!
Yes, this was my concern with this option in https://docs.google.com/document/d/16vhcVeZN2w8iIl_a6gNxd5kFZPPfWh3f5_TnaK9sdHQ

> Since it uses the existing data structure, it’s somewhat complex to add specific metadata for diagnosis. For example, adding a reference to where the name was poisoned doesn’t fit well into this design.

I thought of manipulating `Entry` for that but there is some concern that it makes the structure more complex to use and extend, compared to having a separate set, which might make the lookup more complex.
Since this might change this PR significantly, I think I'll make that a TODO so we can have name poisoning support and improve the diagnostics in a separate PR as suggested.
Added a TODO in `Context::DiagnosePoisonedName()`.